trojsten/rules/kms.py

```diff
@@ -72,22 +78,69 @@ def prepare_coefficients(self, round):
             .annotate(mo_finals=Count("event"))
             .values_list("user", "mo_finals")
         )
-        # We ignore camps that happened before KMS_YEARS_OF_CAMPS_HISTORY years, so we don't
-        # produce too big dictionaries of users.
-        self.camps = dict(
-            EventParticipant.objects.filter(
-                Q(
-                    event__type__name=KMS_CAMP_TYPE,
-                    event__end_time__lt=round.end_time,
-                    event__end_time__year__gte=round.end_time.year - KMS_YEARS_OF_CAMPS_HISTORY,
-                ),
-                Q(going=True) | Q(type=EventParticipant.PARTICIPANT),
-            )
-            .values("user")
-            .annotate(camps=Count("event__semester", distinct=True))
-            .values_list("user", "camps")
+        self.successful_semesters = dict()
+
+        # Get a QuerySet of old semesters. We ignore semesters that happened before
+        # KMS_YEARS_OF_CAMPS_HISTORY years, so we don't produce too big dictionaries of users.
+        old_semesters = Semester.objects.filter(
+            Q(
+                competition=round.semester.competition,
+                year__gte=round.semester.year - KMS_YEARS_OF_CAMPS_HISTORY,
+            ),
+            Q(year__lt=round.semester.year)
+            | Q(
+                year=round.semester.year,
+                number__lt=round.semester.number,
+            ),
         )
```

Comment:
Skusal som sa s tym pohrat a pridat sem `.prefetch_related("event_set", "round_set")` k tomuto query ale nezmenilo velmi ani rychlost ani pocet reads z databazy tak asi to nezmeni vela.